services/mlmodel/client.go

```diff
@@ -41,20 +45,141 @@ func NewClientFromConn(
 	return c, nil
 }
 
-func (c *client) Infer(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error) {
+func (c *client) Infer(ctx context.Context, tensors ml.Tensors, input map[string]interface{}) (ml.Tensors, map[string]interface{}, error) {
 	inProto, err := structpb.NewStruct(input)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+	tensorProto, err := TensorsToProto(tensors)
+	if err != nil {
+		return nil, nil, err
 	}
 	resp, err := c.client.Infer(ctx, &pb.InferRequest{
-		Name:      c.name,
-		InputData: inProto,
+		Name:         c.name,
+		InputTensors: tensorProto,
+		InputData:    inProto,
 	})
 	if err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+	tensorResp, err := protoToTensors(resp.OutputTensors)
+	if err != nil {
+		return nil, nil, err
 	}
+	return tensorResp, resp.OutputData.AsMap(), nil
+}
 
-	return resp.OutputData.AsMap(), nil
+// protoToTensors takes pb.FlatTensors and turns it into a Tensors map.
+func protoToTensors(pbft *pb.FlatTensors) (ml.Tensors, error) {
+	if pbft == nil {
+		return nil, errors.New("protobuf FlatTensors is nil")
+	}
+	tensors := ml.Tensors{}
+	for name, ftproto := range pbft.Tensors {
+		t, err := createNewTensor(ftproto)
+		if err != nil {
+			return nil, err
+		}
+		tensors[name] = t
+	}
+	return tensors, nil
+}
+
+// createNewTensor turns a proto FlatTensor into a *tensor.Dense.
+func createNewTensor(pft *pb.FlatTensor) (*tensor.Dense, error) {
+	shape := make([]int, 0, len(pft.Shape))
+	for _, s := range pft.Shape {
+		shape = append(shape, int(s))
+	}
+	pt := pft.Tensor
+	switch t := pt.(type) {
+	case *pb.FlatTensor_Int8Tensor:
+		data := t.Int8Tensor
+		if data == nil {
+			return nil, errors.New("tensor of type Int8Tensor is nil")
+		}
+		dataSlice := data.GetData()
+		unsafeInt8Slice := *(*[]int8)(unsafe.Pointer(&dataSlice)) //nolint:gosec
+		int8Slice := make([]int8, 0, len(dataSlice))
+		int8Slice = append(int8Slice, unsafeInt8Slice...)
```

Comment:
@acmorrow I think this is "safe" to do, as the protobuf Int8 tensor is actually stored in a "bytes" struct (which is uint8 in Go) so this should just be a reinterpretation of some uint8 bytes into int8 bytes.

Reply:
Yes, that sounds plausible to me.